lightly/api/api_workflow_predictions.py

```diff
@@ -0,0 +1,152 @@
+                Maximum number of workers uploading predictions in parallel.
+
+        Example:
+          >>> import time
+          >>> from tqdm import tqdm
+          >>> from lightly.api import ApiWorkflowClient
+          >>> from lightly.openapi_generated.swagger_client import (
+          >>>     PredictionTaskSchema,
+          >>>     TaskType,
+          >>>     PredictionTaskSchemaCategory,
+          >>> )
+          >>> from lightly.api.prediction_singletons import PredictionSingletonClassificationRepr
+          >>>
+          >>> client = ApiWorkflowClient(
+          >>>     token="MY_LIGHTLY_TOKEN", dataset_id="MY_DATASET_ID"
+          >>> )
+          >>>
+          >>> timestamp = int(time.time())
+          >>> filenames = client.get_filenames()
+          >>> filename_to_prediction_singletons_dummy = {
+          >>>     filename: [PredictionSingletonClassificationRepr(taskName="my-task", categoryId=i%4, score=0.9, probabilities=[0.1, 0.2, 0.3, 0.4])]
+          >>>     for i, filename in enumerate(filenames)
+          >>> }
+          >>> client.create_or_update_predictions(
+          >>>     filename_to_prediction_singletons_dummy,
+          >>>     prediction_uuid_timestamp=timestamp,
+          >>>     progress_bar=tqdm(desc="Uploading predictions", total=len(filenames), unit=" predictions")
+          >>> )
+
+
+        """
+
+        samples = retry(
+            self._samples_api.get_samples_partial_by_dataset_id,
+            dataset_id=self.dataset_id,
+            mode=SamplePartialMode.FILENAMES,
+        )
+        api_filename_to_sample_id = {sample.file_name: sample.id for sample in samples}
+
+        if any(filename_to_upload not in api_filename_to_sample_id for filename_to_upload in filename_to_prediction_singletons.keys()):
+            raise ValueError(f"Some filenames to upload are not found as samples on the server.")
+
+        def upload_prediction(filename_and_predictions_tuple) -> bool:
+            (filename, predictions) = filename_and_predictions_tuple
+            sample_id = api_filename_to_sample_id[filename]
+            prediction_singletons_for_sending = [vars(singleton) for singleton in predictions]
+            retry(
+                self._predictions_api.create_or_update_prediction_by_sample_id,
+                body=prediction_singletons_for_sending,
+                dataset_id=self.dataset_id,
+                sample_id=sample_id,
+                prediction_uuid_timestamp=prediction_uuid_timestamp,
+            )
+
+        # handle the case where len(filename_to_sample_id) < max_workers
+        max_workers = min(len(api_filename_to_sample_id), max_workers)
+        max_workers = max(max_workers, 1)
+
+        with ThreadPoolExecutor(max_workers=max_workers) as executor:
+            for _ in executor.map(upload_prediction, filename_to_prediction_singletons.items()):
+                if progress_bar is not None:
+                    progress_bar.update(1)
```

Comment:
Question: I would expect `filename_to_prediction_singletons.items()` just wrapped in `tqdm`. Why do we accept `progress_bar` as a parameter? To customise the message?

Reply:
Yes, so that we can use `logging_tqdm` instead of `tqdm` when calling it from the worker.

Reply:
We also use it elsewhere, see e.g. https://github.com/lightly-ai/lightly/blob/master/lightly/api/api_workflow_datasources.py#L108